src/hotspot/share/classfile/symbolTable.cpp

```diff
@@ -126,39 +126,78 @@ static uintx hash_shared_symbol(const char* s, int len) {
 #endif
 
 class SymbolTableConfig : public AllStatic {
-private:
+
 public:
-  typedef Symbol* Value;  // value of the Node in the hashtable
+  typedef Symbol Value;  // value of the Node in the hashtable
 
   static uintx get_hash(Value const& value, bool* is_dead) {
-    *is_dead = (value->refcount() == 0);
+    *is_dead = (value.refcount() == 0);
     if (*is_dead) {
       return 0;
     } else {
-      return hash_symbol((const char*)value->bytes(), value->utf8_length(), _alt_hash);
+      return hash_symbol((const char*)value.bytes(), value.utf8_length(), _alt_hash);
     }
   }
   // We use default allocation/deallocation but counted
   static void* allocate_node(void* context, size_t size, Value const& value) {
     SymbolTable::item_added();
-    return AllocateHeap(size, mtSymbol);
+    return allocate_node_impl(size, value);
   }
-  static void free_node(void* context, void* memory, Value const& value) {
+  static void free_node(void* context, void* memory, Value & value) {
     // We get here because #1 some threads lost a race to insert a newly created Symbol
     // or #2 we're cleaning up unused symbol.
     // If #1, then the symbol can be either permanent,
     // or regular newly created one (refcount==1)
     // If #2, then the symbol is dead (refcount==0)
-    assert(value->is_permanent() || (value->refcount() == 1) || (value->refcount() == 0),
-           "refcount %d", value->refcount());
-    if (value->refcount() == 1) {
-      value->decrement_refcount();
-      assert(value->refcount() == 0, "expected dead symbol");
+    assert(value.is_permanent() || (value.refcount() == 1) || (value.refcount() == 0),
+           "refcount %d", value.refcount());
+    if (value.refcount() == 1) {
+      value.decrement_refcount();
+      assert(value.refcount() == 0, "expected dead symbol");
+    }
+    if (value.refcount() != PERM_REFCOUNT) {
+      FreeHeap(memory);
+    } else {
+      MutexLocker ml(SymbolArena_lock, Mutex::_no_safepoint_check_flag); // Protect arena
+      // Deleting permanent symbol should not occur very often (insert race condition),
+      // so log it.
+      log_trace_symboltable_helper(&value, "Freeing permanent symbol");
+      size_t alloc_size = _local_table->get_node_size() + value.byte_size() + value.effective_length();
+      if (!SymbolTable::arena()->Afree(memory, alloc_size)) {
+        log_trace_symboltable_helper(&value, "Leaked permanent symbol");
+      }
```

Comment:
We should avoid freeing if DumpSharedSpaces is true, so that the code is symmetrical with allocate_node_impl(). Maybe this:

```
#if INCLUDE_CDS
  if (DumpSharedSpaces) {
    // no deallocation is needed
  } else
#endif
  if (value.refcount() != PERM_REFCOUNT) {  
```

Reply:
How about performing the above check at the beginning of the function, right after the assert statement as follows?
Let me know if the added assert is necessary.
```
--- a/src/hotspot/share/classfile/symbolTable.cpp
+++ b/src/hotspot/share/classfile/symbolTable.cpp
@@ -151,6 +151,13 @@ public:
     // If #2, then the symbol is dead (refcount==0)
     assert(value.is_permanent() || (value.refcount() == 1) || (value.refcount() == 0),
            "refcount %d", value.refcount());
+#if INCLUDE_CDS
+    if (DumpSharedSpaces) {
+      assert(value.is_permanent(), "symbol should be permanent during CDS dumping");
+      // no deallocation is needed
+      return;
+    }
+#endif
     if (value.refcount() == 1) {
       value.decrement_refcount();
       assert(value.refcount() == 0, "expected dead symbol");
```

Reply:
Ok. I've pushed a commit without the assert statement.